src/Psl/Collection/SetInterface.php

```diff
@@ -0,0 +1,329 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Collection;
+
+use Closure;
+
+/**
+ * @template T of array-key
```

Comment:
the problem is that we do not have a simple ( and more importantly fast ) way to determine what is unique.

I don't know how `Set` in js determines what is unique ( it looks like all `{}` are present, so not sure if it even does so? ), in [rust](https://doc.rust-lang.org/std/collections/struct.HashSet.html), this is done by requiring all elements to implement `Hash` https://doc.rust-lang.org/std/hash/trait.Hash.html and calculating+comparing the hash of the element.

we have 3 ways to do this in psl.

1. support only array-key, this way we can simply make `Set<T>` `array<T, T>` internally, and have PHP handle "uniqueness" for us ( by far the fastest implementation )
2. support only scalars, and use `Psl\Vec\unique_scalar`: this will make the internal table `list<T>`, it is slower than the first solution, but still fast.
3. support all types, and use `Psl\Vec\unique`: this is the slowest, and because of how slow `Psl\Vec\unique` was, we introduced `Psl\Vec\unique_scalar` a while back.

Reply:
In JavaScript, objects are compared per reference and not by value.
It is using the SameValueZero comparison:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#comparing_equality_methods

So I suppose it is using something like Vec\unique internally - but only when appending items. So the slowness is in adding, not in the other methods. I don't really know how slow that acutally is and how many items you are putting in your Set, but in my cases the amount of items in a set is not considered a lot meaning it won't be very slow. But thats depending on the amount of items.

If you are aiming on big data sets, something like a unique hash might be the fastest indeed.
We could create some internal function that does something like this to generate a unique ID:

```php
assert(hash('1') === 'string<1>');
assert(hash(1) === 'int<1>');
assert(hash($object) === 'object<spl_object_id()>');
assert(hash($array) === 'array<recursive_md5()>'); // not sure what to do here

// ...
```

Reply:
I think we should keep it as is for now, the use case i personally have for it does not require support for anything beyond int|string, we could add support for other types later on in v3. but as mentioned, i plan to remove the current `Psl\Collection` component in v4 and re-write it from scratch to align more with Rust `std::collections`.